text/3334-niche.md

```diff
@@ -0,0 +1,348 @@
+possibilities" section). This RFC defines a *simple* mechanism for defining one
+common type of niche, while leaving room for future extension.
+
+Defining a niche mechanism allows libraries to build arbitrary types containing
+niches, and simplifies handling of space-efficient data structures in Rust
+without manual bit-twiddling.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a structure containing exactly one field of a non-zero-sized type
+(non-ZST), you can attach a `niche` attribute on it to declare a specific value
+or range of values for that field as invalid. This promises the compiler that
+you will never store those values in that field, which allows the compiler to
+use those in-memory representations for different purposes, such as the
+representation of `None` in a containing `Option`.
+
+```rust
+use std::mem::size_of;
+
+#[niche(value = 42)]
+struct MeaninglessNumber(u64);
+
+assert_eq!(size_of::<MeaninglessNumber>(), 8);
+assert_eq!(size_of::<Option<MeaninglessNumber>>(), 8);
+
+#[niche(range = 2..)]
+struct Bit(u8);
+
+assert_eq!(size_of::<Bit>(), 1);
+assert_eq!(size_of::<Option<Option<Option<Bit>>>>(), 1);
+```
+
+Constructing a structure with a niche value, or writing to the non-ZST field of
+such a structure, or obtaining a mutable reference to such a field, requires
+`unsafe` code. Causing a type with a niche to contain an invalid value (whether
+by construction, writing, or transmuting) results in undefined behavior.
+
+If a type `T` contains only a single niche value, `Option<T>` (and other enums
+isomorphic to it, with one variant containing `T` and one nullary variant) will
+use that value to represent `None` (the nullary variant). If such a `T` is
+additionally `repr(transparent)` or `repr(C)` or otherwise permitted in FFI,
+`Option<T>` will likewise be permitted in FFI, with the niche value mapping
+bidirectionally to `None` across the FFI boundary.
+
+If a type contains multiple niche values, Rust does not guarantee any
+particular mapping at this time, but may in the future.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The niche attribute may either contain `value = N` where `N` is an unsigned
+integer, or `range = R` where R is a range expression whose endpoints are both
+unsigned integers. The unsigned integers may use any integer base
+representation (decimal, hex, binary, octal), but must not have a type suffix.
+The unsigned integers are interpreted as the bit patterns in memory
+corresponding to the representation of the non-ZST field. For instance, a
+struct with a float field could specify one or more NaN values as a niche using
+the integer representation of those values.
+
+The attribute `#[niche]` may only appear on a struct declaration. The struct
+must contain exactly one field of a non-zero-sized type (non-ZST). The struct
+may contain zero or more ZST fields, such as `PhantomData`. (Note that
```

Comment:
In the first comment you write:

> Define them as narrowly as possible

IMO "as narrowly as possible" would mean "must contain exactly one field" (no ZSTs). This would also allow simplifying the rule about generics to "a struct with the `#[niche]` attribute must not have generic parameters". (I see that this is also mentioned in "unresolved questions").

Not allowing ZST fields would not change the use cases supported by this RPC, I think, because it should be possible to have one struct with the `#[niche]` attribute contained in another `#[repr(transparent)]` wrapper struct.

Reply:
I've now dropped support for ZSTs in the RFC, and moved them to the future possibilities section.